core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -112,12 +112,13 @@ class LogSegment(val log: FileMessageSet,
    * @param startOffset A lower bound on the first offset to include in the message set we read
    * @param maxSize The maximum number of bytes to include in the message set we read
    * @param maxOffset An optional maximum offset for the message set we read
+   * @param maxPosition The maximum position in the log segment that should be exposed for read.
    * 
    * @return The fetched data and the offset metadata of the first message whose offset is >= startOffset,
    *         or null if the startOffset is larger than the largest offset in this log
    */
   @threadsafe
-  def read(startOffset: Long, maxOffset: Option[Long], maxSize: Int): FetchDataInfo = {
+  def read(startOffset: Long, maxOffset: Option[Long], maxSize: Int, maxPosition: Option[Long] = None): FetchDataInfo = {
```

Comment:
@ijuma I haven't thought about this through. My gut feeling here is that when we try to read from the log, it is a bit confusing that we need to provide a capped disk offset we want to read up to. It is only used to prevent the race condition but in most cases it should be none.

So technically I think we should have a default value although there are only two usage of this method.